Fix various C++ coding issues

mario6829 opened the pull request:
Several C++ issues (mainly regarding pointers and character strings, plus a small bug fix) in the code describing ITSU geometry were tackled. New versions of V3Layer.cxx and Detector.cxx/.h were committed. More details in the comments below.

alibuild commented:
Error while checking build/O2/o2-dev-fairroot for b4dc2eecbf3ea65a5216271bb9ef77ae5924c2c0:
```
sw/BUILD/O2-latest/log
[  3%] Building CXX object Common/MathUtils/CMakeFiles/MathUtils.dir/src/Chebyshev3D.cxx.o
[  3%] Linking CXX shared library ../../lib/libHeaders.so
/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/GCC-Toolchain/v6.2.0-alice1-3/bin/../lib/gcc/x86_64-unknown-linux-gnu/6.2.0/../../../../x86_64-unknown-linux-gnu/bin/ld: cannot find -lBoost::signals
collect2: error: ld returned 1 exit status
make[2]: *** [lib/libHeaders.so] Error 1
make[2]: Target `DataFormats/Headers/CMakeFiles/Headers.dir/build' not remade because of errors.
make[1]: *** [DataFormats/Headers/CMakeFiles/Headers.dir/all] Error 2
--
[  5%] Built target ExampleModule1
[  5%] Linking CXX shared library ../../../lib/libFITBase.so
/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/GCC-Toolchain/v6.2.0-alice1-3/bin/../lib/gcc/x86_64-unknown-linux-gnu/6.2.0/../../../../x86_64-unknown-linux-gnu/bin/ld: cannot find -lBoost::signals
collect2: error: ld returned 1 exit status
make[2]: *** [lib/libFITBase.so] Error 1
make[2]: Target `Detectors/FIT/base/CMakeFiles/FITBase.dir/build' not remade because of errors.
make[1]: *** [Detectors/FIT/base/CMakeFiles/FITBase.dir/all] Error 2
--
[  6%] Building CXX object Utilities/QC/QCMerger/CMakeFiles/QCMerger.dir/src/Merger.cxx.o
[  7%] Linking CXX shared library ../../lib/libMathUtils.so
/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/GCC-Toolchain/v6.2.0-alice1-3/bin/../lib/gcc/x86_64-unknown-linux-gnu/6.2.0/../../../../x86_64-unknown-linux-gnu/bin/ld: cannot find -lBoost::signals
collect2: error: ld returned 1 exit status
make[2]: *** [lib/libMathUtils.so] Error 1
make[2]: Target `Common/MathUtils/CMakeFiles/MathUtils.dir/build' not remade because of errors.
make[1]: *** [Common/MathUtils/CMakeFiles/MathUtils.dir/all] Error 2
--
[  8%] Building CXX object CCDB/CMakeFiles/CCDB.dir/src/Backend.cxx.o
[  8%] Linking CXX shared library ../../../lib/libQCViewer.so
/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/GCC-Toolchain/v6.2.0-alice1-3/bin/../lib/gcc/x86_64-unknown-linux-gnu/6.2.0/../../../../x86_64-unknown-linux-gnu/bin/ld: cannot find -lBoost::signals
collect2: error: ld returned 1 exit status
make[2]: *** [lib/libQCViewer.so] Error 1
make[2]: Target `Utilities/QC/QCViewer/CMakeFiles/QCViewer.dir/build' not remade because of errors.
make[1]: *** [Utilities/QC/QCViewer/CMakeFiles/QCViewer.dir/all] Error 2
--
[  9%] Linking CXX shared library ../../../lib/libQCProducer.so
[  9%] Linking CXX shared library ../../../lib/libQCMerger.so
/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/GCC-Toolchain/v6.2.0-alice1-3/bin/../lib/gcc/x86_64-unknown-linux-gnu/6.2.0/../../../../x86_64-unknown-linux-gnu/bin/ld: cannot find -lBoost::signals
collect2: error: ld returned 1 exit status
make[2]: *** [lib/libQCProducer.so] Error 1
make[2]: Target `Utilities/QC/QCProducer/CMakeFiles/QCProducer.dir/build' not remade because of errors.
make[1]: *** [Utilities/QC/QCProducer/CMakeFiles/QCProducer.dir/all] Error 2
[  9%] Building CXX object CCDB/CMakeFiles/CCDB.dir/src/Condition.cxx.o
/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/GCC-Toolchain/v6.2.0-alice1-3/bin/../lib/gcc/x86_64-unknown-linux-gnu/6.2.0/../../../../x86_64-unknown-linux-gnu/bin/ld: cannot find -lBoost::signals
collect2: error: ld returned 1 exit status
make[2]: *** [lib/libQCMerger.so] Error 1
make[2]: Target `Utilities/QC/QCMerger/CMakeFiles/QCMerger.dir/build' not remade because of errors.
make[1]: *** [Utilities/QC/QCMerger/CMakeFiles/QCMerger.dir/all] Error 2
--
[ 11%] Building CXX object Utilities/DataCompression/CMakeFiles/test_DataCompression_test_dc_primitives.dir/test/test_dc_primitives.cxx.o
[ 11%] Linking CXX shared library ../../../lib/libQCMetricsExtractor.so
/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/GCC-Toolchain/v6.2.0-alice1-3/bin/../lib/gcc/x86_64-unknown-linux-gnu/6.2.0/../../../../x86_64-unknown-linux-gnu/bin/ld: cannot find -lBoost::signals
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/710/b4dc2eecbf3ea65a5216271bb9ef77ae5924c2c0/fullLog.txt).

mario6829 commented:
Hi @sawenzel @matthiasrichter , the code submitted in this PR should address the issues you pointed out. Namely

- the **pointer in Detector.h: as @iouribelikov  pointed out in a comment to PR #383, actually there were many raw pointers using new/delete. This had historical reasons, since at the beginning the number of layers was entered by the user at run time. Now this parameter and other similar ones are definitely fixed. So I changed the corresponding member into a constant and all relevant raw pointers into plain arrays in Detector.h ; the code was correspondingly modified, and simplified, in Detector.cxx Please let me know if this is a good way to solve this issue

- use of a pointer cast into V3layer.cxx : I hope I chose the correct way to solve this issue; @matthiasrichter pointed out one example in my last PR (#662), but there were many more similar ones: I addressed all of them, please let me know if I did it in a proper way

- character strings: @matthiasrichter proposed to either change char[] into std:string or use constants for their size; I preferred the second solution since I thought the overhead introduced with string was a bit wasted here. Please take into account that the strings returned by the various methods are now fixed and won't change; the longest one is far below the used size (21 chars vs 30) so there is enough space. Anyway if you still think std:string is better I'll go and change them

- an unrelated small bug was fixed in the copy constructor and in the = operator of the Detector class

Please note that I fixed only those parts I introduced or modified by myself, let me know if there is more I can do to improve the code.
Kindest regards
Mario

alibuild commented:
Error while checking build/O2/o2 for b4dc2eecbf3ea65a5216271bb9ef77ae5924c2c0:
```
sw/BUILD/O2-latest/log
  3/135 Test  #51: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/macro/build_geometry.C ...........................................***Failed    9.44 sec

In file included from input_line_8:1:
/mnt/mesos/sandbox/sandbox/O2/macro/build_geometry.C:40:32: error: no member named 'SimConfig' in namespace 'o2::conf'
  auto& modulelist = o2::conf::SimConfig::Instance().getActiveDetectors();
                     ~~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/build_geometry.C:49:21: error: unknown type name 'FairRunSim'
void build_geometry(FairRunSim* run = nullptr)
                    ^
/mnt/mesos/sandbox/sandbox/O2/macro/build_geometry.C:64:15: error: unknown type name 'FairRunSim'
    run = new FairRunSim();
              ^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/1d2f88e4725d159098f6c65643344202e2cd3790/O2/lib/libField.rootmap declaring 'namespace o2 { namespace field {  } }' were encountered.
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/build_geometry.C:72:31: error: no type named 'MagneticField' in namespace 'o2::field'
  auto field = new o2::field::MagneticField("Maps", "Maps", -1., -1., o2::field::MagFieldParam::k5kG);
                   ~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/1d2f88e4725d159098f6c65643344202e2cd3790/O2/lib/libDetectorsPassive.rootmap declaring 'namespace o2 { namespace passive {  } }namespace o2 { namespace Passive {  } }' were encountered.
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/build_geometry.C:77:16: error: no type named 'Cave' in namespace 'o2::Passive'
  o2::Passive::Cave* cave = new o2::Passive::Cave("CAVE");
  ~~~~~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/build_geometry.C:77:46: error: no type named 'Cave' in namespace 'o2::Passive'
  o2::Passive::Cave* cave = new o2::Passive::Cave("CAVE");
                                ~~~~~~~~~~~~~^
Error in <TInterpreter::AutoParse>: Error parsing payload code for class o2::passive::Hall with content:
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/build_geometry.C:84:34: error: no type named 'Hall' in namespace 'o2::passive'
    auto hall = new o2::passive::Hall("Hall", "Experimental Hall");
                    ~~~~~~~~~~~~~^
Error in <TInterpreter::AutoParse>: Error parsing payload code for class o2::passive::Magnet with content:
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/build_geometry.C:91:36: error: no type named 'Magnet' in namespace 'o2::passive'
    auto magnet = new o2::passive::Magnet("Magnet", "L3 Magnet");
                      ~~~~~~~~~~~~~^
Error in <TInterpreter::AutoParse>: Error parsing payload code for class o2::passive::Dipole with content:
--
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/710/b4dc2eecbf3ea65a5216271bb9ef77ae5924c2c0/fullLog.txt).

alibuild commented:
Error while checking build/o2/macos for b4dc2eecbf3ea65a5216271bb9ef77ae5924c2c0:
```
sw/BUILD/O2-latest/log
 57/135 Test  #51: /build/ci_checks/o2_checker_macos_o2_ci_0/sw/SOURCES/O2/710/0/macro/build_geometry.C ...........................................***Failed   18.88 sec

In file included from input_line_8:1:
/build/ci_checks/o2_checker_macos_o2_ci_0/O2/macro/build_geometry.C:40:32: error: no member named 'SimConfig' in namespace 'o2::conf'
  auto& modulelist = o2::conf::SimConfig::Instance().getActiveDetectors();
                     ~~~~~~~~~~^
--
  auto& modulelist = o2::conf::SimConfig::Instance().getActiveDetectors();
                     ~~~~~~~~~~^
/build/ci_checks/o2_checker_macos_o2_ci_0/O2/macro/build_geometry.C:49:21: error: unknown type name 'FairRunSim'
void build_geometry(FairRunSim* run = nullptr)
                    ^
--
void build_geometry(FairRunSim* run = nullptr)
                    ^
/build/ci_checks/o2_checker_macos_o2_ci_0/O2/macro/build_geometry.C:64:15: error: unknown type name 'FairRunSim'
    run = new FairRunSim();
              ^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /build/ci_checks/o2_checker_macos_o2_ci_0/sw/BUILD/ee78e3f03f6b6c88a041e347e473cd8edd1af273/O2/lib/libField.rootmap declaring 'namespace o2 { namespace field {  } }' were encountered.
--
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/build/ci_checks/o2_checker_macos_o2_ci_0/O2/macro/build_geometry.C:72:31: error: no type named 'MagneticField' in namespace 'o2::field'
  auto field = new o2::field::MagneticField("Maps", "Maps", -1., -1., o2::field::MagFieldParam::k5kG);
                   ~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /build/ci_checks/o2_checker_macos_o2_ci_0/sw/BUILD/ee78e3f03f6b6c88a041e347e473cd8edd1af273/O2/lib/libDetectorsPassive.rootmap declaring 'namespace o2 { namespace passive {  } }namespace o2 { namespace Passive {  } }' were encountered.
--
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/build/ci_checks/o2_checker_macos_o2_ci_0/O2/macro/build_geometry.C:77:16: error: no type named 'Cave' in namespace 'o2::Passive'
  o2::Passive::Cave* cave = new o2::Passive::Cave("CAVE");
  ~~~~~~~~~~~~~^
--
  o2::Passive::Cave* cave = new o2::Passive::Cave("CAVE");
  ~~~~~~~~~~~~~^
/build/ci_checks/o2_checker_macos_o2_ci_0/O2/macro/build_geometry.C:77:46: error: no type named 'Cave' in namespace 'o2::Passive'
  o2::Passive::Cave* cave = new o2::Passive::Cave("CAVE");
                                ~~~~~~~~~~~~~^
Error in <TInterpreter::AutoParse>: Error parsing payload code for class o2::passive::Hall with content:
--
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/build/ci_checks/o2_checker_macos_o2_ci_0/O2/macro/build_geometry.C:84:34: error: no type named 'Hall' in namespace 'o2::passive'
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/710/b4dc2eecbf3ea65a5216271bb9ef77ae5924c2c0/fullLog.txt).

mkrzewic commented:
about returning strings: since you seem to have some "static" strings somewhere, wherever they are defined and in principle only return references to them - why not use gsl::string_span

sawenzel commented:
Changes are reasonable. Merging. Other improvements can be done later (we should try to do a  string_view improvement globally).
